### PR TITLE
CVS-42898/40669/38027 DAG validation improvements

### PR DESCRIPTION
**Pull request by dkalinowski**
https://jira.devtools.intel.com/browse/CVS-42898
https://jira.devtools.intel.com/browse/CVS-40669
https://jira.devtools.intel.com/browse/CVS-38027

Validation fixes/improvements:
- check shape and precision of connections between DLmodel nodes (must match) + tests (CVS-38027)
- for DLmodel nodes, check if all underlying model inputs are connected to some source + tests (CVS-42898)
- check if DLmodel input described in node configuration is present in underlying model instance (previously we only checked outputs) (CVS-40669)
- add parsing pipeline inputs (previously we based on connections from request node)
- adjust unit tests to match new validation methods

This commit also fixes bugs:
https://jira.devtools.intel.com/browse/CVS-42847
https://jira.devtools.intel.com/browse/CVS-42909

### CODE REVIEW

File: src/modelmanager.cpp [hunk 1]
```diff
@@ -167,6 +175,7 @@ void processPipelineConfig(rapidjson::Document& configJson, const rapidjson::Val
 
     std::vector<NodeInfo> info{
         {NodeKind::ENTRY, "request"}};
+    processPipelineInputs(pipelineConfig.FindMember("inputs"), "request", info[0].outputNameAliases);
```
Comment (atobiszei) [inline src/modelmanager.cpp:178]: I think that maybe it would be good to have that name explicit as a static PipelineDefinition string?

File: src/modelmanager.cpp [hunk 2]
```diff
@@ -150,6 +150,14 @@ void processNodeInputs(const std::string nodeName, const rapidjson::Value::Const
     }
 }
 
+void processPipelineInputs(const rapidjson::Value::ConstMemberIterator& pipelineInputsPtr, const std::string& nodeName, std::unordered_map<std::string, std::string>& nodeOutputNameAlias) {
+    for (const auto& pipelineInput : pipelineInputsPtr->value.GetArray()) {
+        SPDLOG_INFO("Alliasing node:{} output:{}, under alias:{}",
+            nodeName, pipelineInput.GetString(), pipelineInput.GetString());
+        nodeOutputNameAlias[pipelineInput.GetString()] = pipelineInput.GetString();
```
Comment (atobiszei) [inline src/modelmanager.cpp:157]: are we sure here we won't get double time the pipelineInput.GetString()?

File: src/pipelinedefinition.cpp [hunk 3]
```diff
@@ -223,91 +223,214 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+                    dependantNodeInfo.nodeName,
+                    dependantNodeInfo.modelName);
                 return StatusCode::FORBIDDEN_MODEL_DYNAMIC_PARAMETER;
             }
         }
 
-        nodeInputs = nodeModelInstance->getInputsInfo();
+        // Save set of inputs which are required by underlying model of currently validated node.
+        // This is later used to make sure we feed each input exactly one data source.
+        std::transform(
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:273]: To check

File: src/pipelinedefinition.cpp [hunk 4]
```diff
@@ -223,91 +223,214 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+    std::unique_ptr<ModelInstanceUnloadGuard> dependantModelUnloadGuard;
+    std::shared_ptr<ModelInstance> dependantModelInstance;
+    std::set<std::string> remainingUnconnectedDependantModelInputs;
+    SPDLOG_DEBUG("Validation of pipeline: {}; node name: {}; node kind: {}",
+        getName(),
+        dependantNodeInfo.nodeName,
+        dependantNodeInfo.kind);
+
+    // If we validate DL model node, retrieve underlying model instance.
+    if (dependantNodeInfo.kind == NodeKind::DL) {
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:236]: Maybe separate method for validation of specific thigs for NodeKind::DL?

This method is already long and hard to comprehend without deep focus.
I am looking for ways how we could encapsulate how much we can.

File: src/pipelinedefinition.cpp [hunk 5]
```diff
@@ -223,91 +223,214 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
-                SPDLOG_ERROR("Validation of pipeline({}) definition failed. Node name {} used model name {} with dynamic shape which is forbidden.", this->pipelineName, node.nodeName, node.modelName);
+        for (const auto& [name, info] : config.getShapes()) {
+            if (info.shapeMode == Mode::AUTO) {
+                SPDLOG_ERROR("Validation of pipeline({}) definition failed. Node name {} used model name {} with dynamic shape which is forbidden.",
+                    this->pipelineName,
+                    dependantNodeInfo.nodeName,
+                    dependantNodeInfo.modelName);
                 return StatusCode::FORBIDDEN_MODEL_DYNAMIC_PARAMETER;
             }
         }
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:269]: checkForForbiddenSettings?/ ensureNotDynamicParametersModel?

**Review by atobiszei — CHANGES_REQUESTED** [review]

File: src/pipelinedefinition.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:236]: This would create bunch of new methods and all would need to have plenty of variables passed over like: unload guard, underlying model instance, current nodeinfo, source node info and connection map. I've tried to comment each step, maybe it is enough?

File: src/modelmanager.cpp (hunk 1 above)
Comment (atobiszei) [reply src/modelmanager.cpp:178]: I mean the names of entry/exit nodes

File: src/modelmanager.cpp (hunk 2 above)
Comment (atobiszei) [reply src/modelmanager.cpp:157]: Possibly we can override value here and not return error?

File: src/test/ensemble_mapping_config_tests.cpp [hunk 6]
```diff
@@ -40,7 +40,7 @@ class PipelineWithInputOutputNameMappedModel : public ::testing::Test {
             ::testing::UnitTest::GetInstance()->current_test_info();
 
         const std::string directoryName = std::string(test_info->test_suite_name());
-        directoryPath = "/tmp" + directoryName;
+        directoryPath = "/tmp/" + directoryName;
```
Comment (atobiszei) [inline src/test/ensemble_mapping_config_tests.cpp:43]: Why this change? I mean tests were silently passing because of that?

File: src/test/ensemble_mapping_config_tests.cpp [hunk 7]
```diff
@@ -250,17 +249,18 @@ TEST_F(PipelineWithInputOutputNameMappedModel, SuccessfullyReferToMappedNamesAnd
     EXPECT_EQ(response_tensor_name->getPrecision(), InferenceEngine::Precision::FP32);
 }
 
-TEST_F(PipelineWithInputOutputNameMappedModel, SuccessfullyReloadPipelineAfterAddingModelMapping) {
+// Disabled until CVS-41658 is resolved.
+TEST_F(PipelineWithInputOutputNameMappedModel, DISABLED_SuccessfullyReloadPipelineAfterAddingModelMapping) {
```
Comment (atobiszei) [inline src/test/ensemble_mapping_config_tests.cpp:253]: This is still not working?

File: src/test/ensemble_metadata_test.cpp [hunk 8]
```diff
@@ -83,20 +83,23 @@ TEST(EnsembleMetadata, MultipleNodesOnDifferentLevelsUsingTheSamePipelineInputs)
     ModelConfig sum_model_config = SUM_MODEL_CONFIG;
     ASSERT_EQ(manager.reloadModelWithVersions(sum_model_config), StatusCode::OK);
 
+    const std::string& INCREMENT_MODEL_INPUT_NAME = DUMMY_MODEL_INPUT_NAME;
```
Comment (atobiszei) [inline src/test/ensemble_metadata_test.cpp:86]: Aren't those defined in test_utils.hpp?

File: src/test/ensemble_tests.cpp [hunk 9]
```diff
@@ -636,8 +636,8 @@ TEST_F(EnsembleFlowTest, CorrectPipelineDefinitionNodesValidation) {
 
     // Simulate reading from pipeline_config.json
     std::vector<NodeInfo> info{
-        {NodeKind::ENTRY, "request"},
-        {NodeKind::DL, "dummy_node", "dummy"},
+        {NodeKind::ENTRY, "request", "", std::nullopt, {{customPipelineInputName, customPipelineInputName}}},
```
Comment (atobiszei) [inline src/test/ensemble_tests.cpp:639]: Not sure if it is worth it at this point but wouldn't it make more sense now to change the order of arguments for NodeInfo constructor? It seems that we almost always need to fill in the last part(outputNameAliases) but we skip (use defaults) for modelName, modelVersion.

Comment (atobiszei) [reply src/test/ensemble_tests.cpp:639]: Maybe to do after CodeFreeze/Next release,

File: src/test/ensemble_tests.cpp [hunk 10]
```diff
@@ -822,18 +827,123 @@ TEST_F(EnsembleFlowTest, PipelineDefinitionNodesWithNodeInputMissingValidation)
     //                                           /\--------|
     connections["dummy_node"] = {
         {"request", {{customPipelineInputName, DUMMY_MODEL_OUTPUT_NAME}}},
-        {"dummy_node", {{DUMMY_MODEL_OUTPUT_NAME, "MISSING"}}}};
+        {"dummy_node", {{DUMMY_MODEL_OUTPUT_NAME, "MISSING"}}}};  // loop?
```
Comment (atobiszei) [inline src/test/ensemble_tests.cpp:830]: leftover debug comment?

File: src/test/ensemble_tests.cpp [hunk 11]
```diff
@@ -822,18 +827,123 @@ TEST_F(EnsembleFlowTest, PipelineDefinitionNodesWithNodeInputMissingValidation)
...
+
+    connections["response"] = {
+        {"dummy_node_1x20", {{DUMMY_MODEL_OUTPUT_NAME, customPipelineOutputName}}}};
+
+    // Create pipeline definition
+    std::unique_ptr<PipelineDefinition> pipelineDefinition = std::make_unique<PipelineDefinition>("my_new_pipeline", info, connections);
+    ASSERT_EQ(pipelineDefinition->validateNodes(manager), StatusCode::INVALID_SHAPE);
+}
+
+// Disabled until CVS-36446 is done.
```
Comment (atobiszei) [inline src/test/ensemble_tests.cpp:908]: Can we somewhat test it even without support for all precisions? I think I32/FP32 are supported so maybe use only those for test?
My fear is that this code can quickly get outdated if disabled.

File: src/test/ensemble_tests.cpp [hunk 12]
```diff
@@ -822,18 +827,123 @@ TEST_F(EnsembleFlowTest, PipelineDefinitionNodesWithNodeInputMissingValidation)
...
+
+    // Simulate reading from pipeline_config.json
+    std::vector<NodeInfo> info{
+        {NodeKind::ENTRY, "request", "", std::nullopt, {{customPipelineInputName, customPipelineInputName}}},
+        {NodeKind::DL, "dummy_node_fp32", "dummy_fp32", std::nullopt, {{DUMMY_MODEL_OUTPUT_NAME, DUMMY_MODEL_OUTPUT_NAME}}},
+        {NodeKind::DL, "dummy_node_u8", "dummy_u8", std::nullopt, {{DUMMY_MODEL_OUTPUT_NAME, DUMMY_MODEL_OUTPUT_NAME}}},
+        {NodeKind::EXIT, "response"},
+    };
+
+    std::unordered_map<std::string, std::unordered_map<std::string, InputPairs>> connections;
```
Comment (atobiszei) [inline src/test/ensemble_tests.cpp:930]: I think it may be time to chave somewhere typedef/using declared for connections.

Comment (dkalinowski) [reply src/test/ensemble_tests.cpp:930]: We already have it defined. We just not use it for some reason. Will look for all occurrences and replace.

File: src/test/ensemble_mapping_config_tests.cpp (hunk 6 above)
Comment (dkalinowski) [reply src/test/ensemble_mapping_config_tests.cpp:43]: Tests were passing, but files were not created in /tmp directory. Purpose of this test suite was to place it everything under /tmp.

File: src/test/ensemble_mapping_config_tests.cpp (hunk 7 above)
Comment (dkalinowski) [reply src/test/ensemble_mapping_config_tests.cpp:253]: This is not working due to CVS-41658 bug, after reload with new mapping config we get tensor map out of sync.

File: src/test/ensemble_tests.cpp (hunk 11 above)
Comment (dkalinowski) [reply src/test/ensemble_tests.cpp:908]: We would need to add yet another dummy model to unit test. Do we want the same model (dummy) but with other precision just for this one test?

File: src/test/ensemble_metadata_test.cpp (hunk 8 above)
Comment (dkalinowski) [reply src/test/ensemble_metadata_test.cpp:86]: I rename the variable to better match its purpose in this specific unit test.

File: src/modelmanager.cpp (hunk 2 above)
Comment (atobiszei) [reply src/modelmanager.cpp:157]: It seems that during checking that we found few issues in other places as well. You can double the inputs/outputs etc. and there is no warning (I think those values are just overridden.

For me those issues are  to be fixed in separate bug.

Comment (dkalinowski) [reply src/modelmanager.cpp]: Done:  
https://jira.devtools.intel.com/browse/CVS-43541  
https://jira.devtools.intel.com/browse/CVS-43543  
https://jira.devtools.intel.com/browse/CVS-43544

File: src/modelmanager.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/modelmanager.cpp]: Done

File: src/pipelinedefinition.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Separated validation into NodeValidator class. Separated parts of validation code into smaller methods.

File: src/pipelinedefinition.cpp (hunk 5 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Done

File: src/test/ensemble_tests.cpp (hunk 10 above)
Comment (dkalinowski) [reply src/test/ensemble_tests.cpp]: Done

File: src/test/ensemble_tests.cpp (hunk 12 above)
Comment (dkalinowski) [reply src/test/ensemble_tests.cpp]: Done

File: src/modelmanager.cpp [hunk 13]
```diff
@@ -150,6 +150,19 @@ void processNodeInputs(const std::string nodeName, const rapidjson::Value::Const
...
+void processPipelineInputs(const rapidjson::Value::ConstMemberIterator& pipelineInputsPtr, const std::string& nodeName, std::unordered_map<std::string, std::string>& nodeOutputNameAlias, const std::string& pipelineName) {
+    for (const auto& pipelineInput : pipelineInputsPtr->value.GetArray()) {
+        const std::string pipelineInputName = pipelineInput.GetString();
+        SPDLOG_INFO("Alliasing node:{} output:{}, under alias:{}",
+            nodeName, pipelineInputName, pipelineInputName);
+        if (nodeOutputNameAlias.count(pipelineInputName) > 0) {
+            SPDLOG_WARN("Pipeline {} has duplicated input declaration", pipelineName);
+        } else {
+            nodeOutputNameAlias[pipelineInputName] = pipelineInputName;
+        }
```
Comment (atobiszei) [inline src/modelmanager.cpp:162]: minor:
Efficient construction for this would be:
auto it = nodeOutputNameAlias.find(name)
if (it == nodeOutputNameAlias.end()) {
   SPDLOG_ERROR
} else {
    (*it).second = name;
}

WIth current way you are doing search on a map twice.

File: src/test/ensemble_tests.cpp (hunk 11 above)
Comment (atobiszei) [reply src/test/ensemble_tests.cpp:894]: I am good to wait for the fix.

File: src/pipelinedefinition.cpp [hunk 14]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+                    return result;
+                }
+            }
+        }
+
+        return StatusCode::OK;
+    }
+
+    Status validate() {
+        // If we validate DL model node, retrieve underlying model instance.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:471]: Unnecessary comment now?

File: src/pipelinedefinition.cpp [hunk 15]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+
+            result = checkForForbiddenDynamicParameters();
+            if (!result.ok()) {
+                return result;
+            }
+
+            prepareRemainingUnconnectedDependantModelInputsSet();
+        }
+
+        // Check all connections entering currently validated node.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:486]: Unnecessary comment?

File: src/pipelinedefinition.cpp [hunk 16]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+                dependantNodeInfo.modelName,
+                dependantNodeInfo.modelVersion.value_or(0),
+                name);
+            return StatusCode::PIPELINE_MODEL_INPUT_CONNECTED_TO_MULTIPLE_DATA_SOURCES;
+        }
+        return StatusCode::OK;
+    }
+
+    Status validateConnection(const NodeInfo& dependencyNodeInfo, const InputPairs& mapping) {
+        // At this point dependency node can only be either DL model node or entry node.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:424]: Hmm maybe throw an exception otherwise instead of comment?

File: src/pipelinedefinition.cpp [hunk 17]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+        // Take care when adding new node types.
+        // If underlying model is of type DL model, retrieve underlying model instance for later use.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:426]: Not sure if we need to need that comment either.

File: src/pipelinedefinition.cpp [hunk 18]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+        }
 
-            for (auto alias : connection.second) {
-                std::string& dependencyOutputAliasName = alias.first;
-                std::string dependencyOutputName;
-                if (sourceNodeInfo->outputNameAliases.count(dependencyOutputAliasName)) {
-                    dependencyOutputName = sourceNodeInfo->outputNameAliases[dependencyOutputAliasName];
-                } else {
-                    dependencyOutputName = dependencyOutputAliasName;
+        // Validate each connection between dependency and dependant node.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:445]: Same here.

**Review by atobiszei — APPROVED** [review]

File: src/pipelinedefinition.cpp [hunk 19]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+            SPDLOG_ERROR("Validation of pipeline({}) definition failed. Node (name:{}) is connected to missing dependency node (name:{})",
+                pipelineName,
+                dependantNodeInfo.nodeName,
+                dependencyNodeName);
+            return StatusCode::PIPELINE_NODE_REFERING_TO_MISSING_NODE;
+        }
+
+        // Exit cannot be dependency of any node.
+        if (dependencyNodeInfo->kind == NodeKind::EXIT) {
+            return StatusCode::UNKNOWN_ERROR;
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:281]: Don't we have any more telling status code? I am not sure if this would fall out at the stage of json schema validation.

File: src/pipelinedefinition.cpp [hunk 20]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+        }
+
+        return StatusCode::OK;
+    }
 
-        auto& config = nodeModelInstance->getModelConfig();
-        if (config.getBatchingMode() == Mode::AUTO) {
-            SPDLOG_ERROR("Validation of pipeline({}) definition failed. Node name {} used model name {} with dynamic batch size which is forbidden.", this->pipelineName, node.nodeName, node.modelName);
+    Status checkForForbiddenDynamicParameters() {
+        // Ban creating pipelines with Dynamic Model Parameters.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:288]: Unnecessary?

File: src/pipelinedefinition.cpp [hunk 21]
```diff
@@ -216,92 +216,300 @@ void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
...
+                dependantNodeInfo.modelName,
+                dependantNodeInfo.modelVersion.value_or(0),
+                modelInputName,
+                tensorInput->getPrecisionAsString(),
+                dependencyNodeInfo.nodeName,
+                dependencyNodeInfo.modelName,
+                dependencyNodeInfo.modelVersion.value_or(0),
+                modelOutputName,
+                tensorOutput->getPrecisionAsString());
+            return StatusCode::INVALID_PRECISION;  // CANNOT REACH MANUALLY, WE HAVE NO OTHER PRECISION SUPPORT
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:362]: I do not understand the comment here

File: src/modelmanager.cpp [hunk 22]
```diff
@@ -150,6 +150,19 @@ void processNodeInputs(const std::string nodeName, const rapidjson::Value::Const
     }
 }
 
+void processPipelineInputs(const rapidjson::Value::ConstMemberIterator& pipelineInputsPtr, const std::string& nodeName, std::unordered_map<std::string, std::string>& nodeOutputNameAlias, const std::string& pipelineName) {
+    for (const auto& pipelineInput : pipelineInputsPtr->value.GetArray()) {
+        const std::string pipelineInputName = pipelineInput.GetString();
+        SPDLOG_INFO("Alliasing node:{} output:{}, under alias:{}",
```
Comment (dtrawins) [inline src/modelmanager.cpp:156]: Mapping would be better work.
While this function is called for every request, maybe DEBUG model would be more applicable here

File: src/modelmanager.cpp [hunk 23]
```diff
@@ -150,6 +150,19 @@ void processNodeInputs(const std::string nodeName, const rapidjson::Value::Const
...
+            nodeName, pipelineInputName, pipelineInputName);
+        if (nodeOutputNameAlias.count(pipelineInputName) > 0) {
+            SPDLOG_WARN("Pipeline {} has duplicated input declaration", pipelineName);
```
Comment (dtrawins) [inline src/modelmanager.cpp:159]: there is not warning level. Should it be error?

File: src/modelmanager.cpp (hunk 22 above)
Comment (dkalinowski) [reply src/modelmanager.cpp:156]: This is not called for every request, it is called when we read pipeline from config.json file. Most of pipeline add/reload/validate logs are INFO for development purpose, there will be one last commit before code freeze that changes log levels.

File: src/modelmanager.cpp (hunk 23 above)
Comment (dkalinowski) [reply src/modelmanager.cpp:159]: This is related to bug I created: https://jira.devtools.intel.com/browse/CVS-43541  
For now I just added warning if inputs are duplicated. Such pipeline works fine. Ideal solution would be to restrict administrator from creating such pipelines. This is expained in the jira bug.

File: src/modelmanager.cpp (hunk 13 above)
Comment (dkalinowski) [reply src/modelmanager.cpp:162]: The logic here is opposite: we only insert if we cannot find key. Nevertheless we can switch to use .find()

File: src/pipelinedefinition.cpp (hunk 16 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:424]: This comment is to remember in next release of model ensemble to update this part of code.

File: src/pipelinedefinition.cpp (hunk 19 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:281]: I was sure that we cannot reach this part of code due to earlier stages of validation. Just checked and this is possible. Adding new status code and unit test for this scenario.

File: src/modelmanager.cpp (hunk 13 above)
Comment (dkalinowski) [reply src/modelmanager.cpp]: Done, I used .insert(), which returns information if entry was added or skipped due to key duplicatio.

File: src/pipelinedefinition.cpp (hunk 14 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Removed

File: src/pipelinedefinition.cpp (hunk 15 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Removed

File: src/pipelinedefinition.cpp (hunk 17 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Removed

File: src/pipelinedefinition.cpp (hunk 18 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Removed

File: src/pipelinedefinition.cpp (hunk 19 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Done

File: src/pipelinedefinition.cpp (hunk 20 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Removed

File: src/pipelinedefinition.cpp (hunk 21 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Debug leftover, removed

File: src/modelmanager.cpp (hunk 22 above)
Comment (dkalinowski) [reply src/modelmanager.cpp]: Done

File: src/modelmanager.cpp (hunk 23 above)
Comment (dkalinowski) [reply src/modelmanager.cpp]: Done

**Review by dtrawins — APPROVED** [review]

**Review by atobiszei — APPROVED** [review]
